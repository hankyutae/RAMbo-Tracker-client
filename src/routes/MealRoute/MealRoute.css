.AddMealDiv{
  box-sizing: border-box;
  min-width:300px;
  max-width:900px;
  margin:auto;
  margin-top:7vh;
  padding:10px;
  background-color:gainsboro;
}
.AddMealForm{

}
#AddMealNameInput{

}

.react-datepicker__input-container{
  display:inline-block;
}

.InlineBlockTheDate{
  display:inline-block;
}

.AddMealFoodsDisplayContainer{
}
.AddMealFoodsDisplayRow{
  display:flex;
  justify-content:space-between;
  align-items:center;
}
.AddMealFoodsDisplayRow div{
  word-break: normal; 
  word-wrap: normal;
  display:inline-block;
}
.AddMealFoodsDisplayName{
  flex-basis:60%;
  flex-grow:1;
}
.AddMealFoodsDisplaySpace{
  flex-basis:20px;
  flex-grow:1;
}
.AddMealFoodsDisplayRemove{
  flex-grow:0;
  height:25px;
  color:red;
}