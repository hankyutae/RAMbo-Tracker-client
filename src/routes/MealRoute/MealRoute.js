import React from 'react';
import './MealRoute.css';
import DatePicker from "react-datepicker";
import AddFood from '../../components/AddFood/AddFood';
import FoodApiService from '../../services/food-api-service';
import MealApiService from '../../services/meal-api-service';
import ProcessFoodName from '../../services/process-food-name';
import ValidationError from '../../components/ValidationError/ValidationError';
 
import "react-datepicker/dist/react-datepicker.css";

export default class MealRoute extends React.Component{
  state={
    mealName:'A Meal',
    mealTime:new Date(),
    foodsInMeal:[],
    
    handleAddFood:(food)=>{
      FoodApiService.postFood(food.ndbno,food.name)
        .then(res=>{
          console.log(...this.state.foodsInMeal,food);
          this.setState({
            foodsInMeal:[...this.state.foodsInMeal,food]
          });
      })
    }
  }

  handleNameChange = mealName=>{
    this.setState({
      mealName
    });
  }
  handleTimeChange = date => {
    console.log(this.state.mealTime);
    console.log(typeof this.state.mealTime);
    this.setState({
      mealTime: date
    });
  }

  verify= () =>{
    return this.verifyFoodNonempty() || this.verifyMealName();
  }

  verifyMealName= ()=>{
    if(this.state.mealName==='')
      return 'Meal name can\'t be empty';
  }

  verifyFoodNonempty=() =>{
    if(this.state.foodsInMeal.length===0)
      return 'Meal contents can\'t be empty';
  }

  handleRemoveFood=(e,index)=>{
    e.preventDefault();
    e.stopPropagation();
    let foods=[...this.state.foodsInMeal];
    foods.splice(index,1);
    this.setState({
      foodsInMeal:foods
    })
  }

  handleMealSubmit= (e)=>{
    e.preventDefault();
    if( !(this.verifyMealName()) && !(this.verifyFoodNonempty()) ){
      /* console.log({
        time:this.state.mealTime,
        name:this.state.mealName,
        items: this.state.foodsInMeal.map(food=>food.ndbno)
      }); */
      MealApiService.postMeal({
        time:this.state.mealTime,
        name:this.state.mealName,
        items: this.state.foodsInMeal.map(food=>food.ndbno)
      })
        .then(()=>{
          //this.props.history.push('/')
          this.props.closeModal('addMealModal');//this functions is passed in from dashboard to close the modal, it should be placed int the 'then' of api call to ensure it only runs in happy case 
        });
    }
  }
  
  render(){
    return (
      <div className='add-meal-div'>
        <h2>Add a Meal!</h2>
        <form className='add-meal-form' onSubmit={this.handleMealSubmit}>
          <label htmlFor='AddMealNameInput'>
            Meal Name:
          </label>
          <br></br>
          <input 
            type='text' 
            onChange={(e)=>this.handleNameChange(e.target.value)} 
            id='AddMealNameInput'
            value={this.state.mealName}
          />
          <ValidationError message={this.verifyMealName()} />
          <br></br>
          <label htmlFor='AddMealTimeInput'>
            Date:
          </label>
          <DatePicker
            selected={this.state.mealTime}
            onChange={this.handleTimeChange}
            showTimeSelect
            withPortal
            dateFormat="Pp"
          />
          <div className='AddMealFoodsDisplayContainer'>
            <h3>Meal Contents</h3>
            <div className='AddMealFoodsDisplay'>
              {
                this.state.foodsInMeal.map((food,index)=>
                <div key={index}>

                  <div  className='AddMealFoodsDisplayRow'>
                    <div className='AddMealFoodsDisplayName'>
                      {ProcessFoodName(food.name)}
                    </div>
                    <div className='AddMealFoodsDisplaySpace'>
                    </div>
                    <button className="user-button" onClick={e => { this.handleRemoveFood(e,index) }}>
                      Remove
                    </button>
                  </div>
                  <hr></hr>
                </div>
                )
              }
            </div>
          </div>


          <AddFood addFood={this.state.handleAddFood}/>

          <br></br>


          <button disabled={this.verify()? true:false} className='user-button'>
            Add Meal
          </button>
          <ValidationError message={this.verifyFoodNonempty()} />
        </form>
      </div>
    );
  }
}