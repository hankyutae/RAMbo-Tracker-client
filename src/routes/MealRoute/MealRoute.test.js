import React from 'react';
import ReactDOM from 'react-dom';
import MealRoute from './MealRoute';
import { BrowserRouter } from 'react-router-dom';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <BrowserRouter>
      <MealRoute />
    </BrowserRouter>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});