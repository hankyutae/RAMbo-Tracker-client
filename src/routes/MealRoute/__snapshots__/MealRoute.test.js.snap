// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<div
  className="add-meal-div"
>
  <h2>
    Log a Meal
  </h2>
  <form
    className="add-meal-form"
    onSubmit={[Function]}
  >
    <label
      className="add-meal-labels"
      htmlFor="AddMealNameInput"
    >
      Meal Name:
    </label>
    <br />
    <input
      id="AddMealNameInput"
      onChange={[Function]}
      type="text"
      value="A Meal"
    />
    <br />
    <label
      className="add-meal-labels"
      htmlFor="AddMealTimeInput"
    >
      Date and Time:
    </label>
    <div>
      <div
        className="react-datepicker__input-container"
      >
        <input
          className=""
          disabled={false}
          id="AddMealDateInput"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="text"
          value="09/24/2019, 11:47 AM"
        />
      </div>
    </div>
    <div
      className="AddMealFoodsDisplayContainer"
    >
      <h3>
        Meal Contents
      </h3>
      <div
        className="AddMealFoodsDisplay"
      />
    </div>
    <div
      className="AddFoodDiv"
    >
      <label
        htmlFor="searchTermInput"
      >
        Search for a food:
      </label>
      <br />
      <div
        className="SearchFoodForm"
      >
        <input
          form="sub-form"
          id="searchTermInput"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value=""
        />
        <button
          className="user-button"
          form="sub-form"
          htmlFor="searchTermInput"
          id="search-food-button"
          onClick={[Function]}
        >
          Search for Food
        </button>
      </div>
    </div>
    <br />
    <div
      className="ValidationError"
    >
      Meal contents can't be empty
    </div>
    <button
      className="user-button"
      disabled={true}
      id="add-meal-button"
    >
      Add Meal
    </button>
  </form>
</div>
`;
